src/trimerge-ordered-map.ts

```diff
@@ -0,0 +1,127 @@
+import { diff3Keys } from './diff3-keys';
+import { MergeFn } from './trimerge';
+import { Path } from './path';
+import { diffIndices } from './node-diff3';
+
+export function internalTrimergeOrderedMap<K, V>(
+  origMap: Map<K, V>,
+  leftMap: Map<K, V>,
+  rightMap: Map<K, V>,
+  path: Path,
+  mergeFn: MergeFn,
+  allowOrderConflicts: boolean,
+  callback: (key: K, value: V) => void,
+): 'left' | 'right' | undefined {
+  const leftRightKeys = new Set(leftMap.keys());
+  for (const key of rightMap.keys()) {
+    leftRightKeys.add(key);
+  }
+
+  // Merge all values first (might restore some deleted values)
+  const mergedValues = new Map<K, V>();
+  const restoreLeft = new Set<K>();
+  const restoreRight = new Set<K>();
+  for (const key of leftRightKeys) {
+    const left = leftMap.get(key);
+    const right = rightMap.get(key);
+    const merged = mergeFn(
+      origMap.get(key),
+      left,
+      right,
+      [...path, key as any],
+      mergeFn,
+    );
+    if (merged !== undefined) {
+      if (!leftMap.has(key)) {
+        restoreLeft.add(key);
+      }
+      if (!rightMap.has(key)) {
+        restoreRight.add(key);
+      }
+      mergedValues.set(key, merged);
+    }
+  }
+  const orig = Array.from(origMap.keys());
+  const left = restoreKeys(orig, leftMap, restoreLeft);
+  const right = restoreKeys(orig, rightMap, restoreRight);
```

Comment:
The dependence in `restoreKeys` on `restoreLeft` and `restoreRight` feels pretty tightly coupled. Could there be a way to consolidate the logic required to compute `restoreLeft`/`restoreRight` and to call `restoreKeys`? I don't have anything in mind because we're also calling `mergeFn`... just a thought.